cabal-install/Distribution/Client/Setup.hs

```diff
@@ -387,18 +399,96 @@ instance Monoid GlobalFlags where
     where combine field = field a `mappend` field b
 
 withGlobalRepos :: Verbosity -> GlobalFlags -> ([Repo] -> IO a) -> IO a
-withGlobalRepos _verbosity globalFlags callback =
-    callback $ remoteRepos ++ localRepos
+withGlobalRepos verbosity globalFlags callback =
+    withSecureRepos verbosity secureRepos $ \secureRepos' ->
+      callback $ secureRepos' ++ remoteRepos ++ localRepos
   where
+    secureRepos =
+      [ (remote, cacheDir)
+      | remote <- fromNubList $ globalRemoteRepos globalFlags
+      , remoteRepoSecure remote
+      , let cacheDir = fromFlag (globalCacheDir globalFlags)
+                   </> remoteRepoName remote ]
     remoteRepos =
       [ RepoRemote remote cacheDir
       | remote <- fromNubList $ globalRemoteRepos globalFlags
+      , not (remoteRepoSecure remote)
       , let cacheDir = fromFlag (globalCacheDir globalFlags)
                    </> remoteRepoName remote ]
     localRepos =
       [ RepoLocal local
       | local <- fromNubList $ globalLocalRepos globalFlags ]
 
+-- | Initialize the provided secure repositories
+--
+-- Assumed invariant: `remoteRepoSecure` should be set for all these repos.
+withSecureRepos :: forall a.
+                   Verbosity
+                -> [(RemoteRepo, FilePath)]  -- ^ Repos and cacheDirs
+                -> ([Repo] -> IO a)          -- ^ Callback
+                -> IO a
+withSecureRepos verbosity repos callback = go [] repos
+  where
+    go :: [Repo] -> [(RemoteRepo, FilePath)] -> IO a
+    go acc []                = callback (reverse acc)
+    go acc ((r,cacheDir):rs) = do
+      cachePath <- Sec.makeAbsolute $ Sec.fromFilePath cacheDir
+      withSecureRepo verbosity r cachePath $ \r' ->
+        go (RepoSecure r cacheDir r':acc) rs
+
+-- | Initialize the given secure repo
+--
+-- The security library has its own concept of a "local" repository, distinct
```

Comment:
Note that HEAD already has a new kind of section; we now have, "old style" single line `remote-repo` definitions

```
remote-repo: hackage.haskell.org:http://hackage.haskell.org/packages/archive
```

as well as fancy new _sections_ (not yet available in any released version of `cabal-install`, AFAIK)

```
remote-repo hackage.haskell.org
  url: http://hackage.haskell.org/
  secure: True
  root-keys: 1035a452fd3ada87956f9e77595cfd5e41446781d7ba9ff9e58b94488ac0bad7
             89e692e45b53b575f79a02f82fe47359b0b577dec23b45d846d6e734ffcc887a
             dc4b6619e8ea2a0b72cad89a3803382f6acc8beda758be51660b5ce7c15e535b
  key-threshold: 2
```

we could, if we wanted, rename the latter to `repository`, and then slowly phase out both `local-repo` and `remote-repo`. Anyway, I don't feel too strongly about any of these, I'm just pointing out options.